### PR TITLE
Fixes custom entries precedence bugs, including #11331 and a small enhancement

### LINKED ISSUE #11331: constr in Custom Entry mistaken for SELF
#### Description of the problem

The custom entry below treats parts `p` and `b` as `custom expr` instead of the specified `constr`.

```
Declare Custom Entry expr. Declare Custom Entry aux.
Notation "p u b" := (p (fun _ => b) u)
  (in custom expr at level 201, p constr, b constr, u custom aux at level 1).
Print Custom Grammar expr.
(*
Entry constr:expr is
[ "201" RIGHTA
  [ SELF; constr:aux LEVEL "1"; NEXT ] ]
*)
```

I suspect that this breaks some uses of my (more complicated) notations.

#### Coq Version

master, 8.10.2

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix

Fixes / closes #11331 as well as two small custom entries bugs + an enhancement.

The two other bugs are:
- Respect of strictly-less precedences (due to associativity) when looking for coercions.
- Skip levels for custom entries in `ppconstr.ml` (rely instead on `constrextern.ml`).

The enhancement is:
- Bypassing a coercion to constr for printing a global in a custom entry if this entry has a rule for globals..

Five other commits are about restructurations. They should not change the semantics.

There is also what is suspectingly a small precedence bug in printing `Tactic Notation` (but I did not investigate deeply).

- [X] Added / updated test-suite
- [X] Entry added in the changelog (suggestions for better phrasing welcome)

### CODE REVIEW

**Review by maximedenes — CHANGES_REQUESTED** [review]
Please separate improvements and bug fixes (ideally one PR per fix if they are orthogonal). Mixing fixes and improvements is often inconvenient later (for example if we need to revert the improvements, or if we backport only the fixes).

**Comment by maximedenes** [comment]
I'll try to have a look at the fix itself tomorrow or next week, if nobody else does.

**Comment by herbelin** [comment]
> Please separate improvements and bug fixes (ideally one PR per fix if they are orthogonal). Mixing fixes and improvements is often inconvenient later (for example if we need to revert the improvements, or if we backport only the fixes).

Ideally, this PR could be split into 7 different small PRs (3 fixes, one enhancement, 3 cleanups interleaved), but that's a lot! I could do these 7 several PRs or whatever is convenient to help reviewers, depending on their time available. Note that reverting would probably not be more/less difficult with 7 PRs merged in sequence than with one PR made of the same 7 sequential commits merged altogether, but whatever, any sensible solution that would help is ok to me, just tell me what would suit best your willingness to review.

> I'll try to have a look at the fix itself tomorrow or next week, if nobody else does.

I guess you mean _the_ fix for #11331. It is the second commit and it is in two parts: the part already studied by @ppedrot in the discussion for #11331 and another part which raises an error rather than using level 200 when a custom entry which is not at the border of a rule has no indicated level. Thanks in advance!

**Comment by herbelin** [comment]
@maximedenes: If it can help your review, here are some informations about how subentries are treated.

Consider for instance `Reserved Notation "x + y" (at level 50, left associativity).`.

First, informations about `x` and `y` are collected:
- at parsing of the notation itself, it is not the case here, but, in general, the modifiers may say something about `x` and `y`
- still at parsing of the notation, `find_symbols` recognizes that `x` and `y` are at the border and thus sensitive to the `associativity` flag
- still at parsing of the notation, `set_subentry_type` (formerly `set_entry_type`) combines the knowledge about modifiers, the knowledge about being at the border or not, the knowledge of associativity to determine that `x` is `ETConstr (InConstrEntry,None,(50,BorderProd (Left,LeftA)))`

Now, this last information is used in different ways:
- for declaring the rule to camlp5, it is `Egramcoq.symbol_of_target` which applies: it transforms the knowledge about `x` and `y` into camlp5 commands such as `NEXT`, `SELF`, ...; moreover, a heuristics is used to decide when using `SELF` rather than `constr LEVEL n`
- for uniquely identifying the notation, these are turned into relative levels by `Metasyntax.precedence_of_entry_type`
- for declaring the rule to the Coq printer, it is (now) `Metasyntax.unparsing_precedence_of_entry_type`; its particularity is that it discards precedences for custom entries because these are turned into explicit "parsing coercions" (such as `Notation "( x )"  (in custom expr at level 0, x at level 2)`)
- on the opposite side, `subentry_of_constr_prod_entry` is used to deal with the precedences of `custom` entries in `constrextern`: it discards the information about precedences for "constr" but keep the one for custom entries
- for interpretation the right-hand side of the notation, the informations about `x` and `y` are turned into `interpretation type` by `Metasyntax.make_interpretation_type`
- for the purpose of dealing with special class of rules such as parsing coercions, it is `Metasyntax.absolute_entry_of_constr_prod_entry`

So, to summarize, there are different uses, different special cases and I don't know how much this could be simplified further.

**Comment by ejgallego** [comment]
IMHO this is fine for 8.12 and I can assign it; but indeed how to address the things targeted to 8.11 may need splitting in the line Maxime asked, @ppedrot how would you like to see this PR split?

**Comment by ppedrot** [comment]
@herbelin I think that a minimal, separate fix for #11331 targetting 8.11 would be welcome. We're too close to the release to perform unrelated cleanups.

**Comment by herbelin** [comment]
The fixes for #11331 and for skipping levels for custom entries in `ppconstr.ml` (commits 2 and 3), as well as the enhancement (commit 6) can probably be easily isolated .

The third fix (respect of strictly-less precedences (due to associativity) requires some deeper changes and the cleanups are related.

I think the best I could do is to isolate commits 2 and 3 (and possibly 6) for 8.11 and leave the third bug for 8.12. (But I don't think there is a danger to include the cleanup anyway.)

What is your planned schedule?

**Comment by ppedrot** [comment]
@herbelin No explicit schedule, but trying to release asap. With a bit of luck this is achievable by tomorrow.

**Comment by ppedrot** [comment]
@herbelin It would really help to split this PR in atomic patches. I am willing to review the changes provided you turn this into several PRs. Even for after-the-fact analysis of the code (e.g. git bisect), it is always better to have semantically distinct components.

**Comment by herbelin** [comment]
@ppedrot 
> It would really help to split this PR in atomic patches. I am willing to review the changes provided you turn this into several PRs. 

Then, I split this PR into five parts:

- #11530 (which contains the fix to #11331) 
- herbelin/github-coq#5 (which is pure cleanup about the printing data types)
- herbelin/github-coq#6 (which is an improvement in printing abbreviations in custom entries)
- herbelin/github-coq#7 (which fixes a parentheses printing bug in custom entries)
- herbelin/github-coq#8 (which fixes what seems to be a minor printing bug in printing notations)

Note that I experimented pushing the PRs on my own github fork so that the independent delta can be reviewed more easily. When reviews are done, I can move the PR on the main github so that they can be merged (or, maybe, will it be possible to merge directly from my fork?).

Tell me if this fits what you had in mind.

> Even for after-the-fact analysis of the code (e.g. git bisect), it is always better to have semantically distinct components.

I agree with that. I was concerned by the number of reviews it was needing but you are sure that the process will go faster by splitting PRs into more PRs, I'd be happy with it.

The current PR can now be closed.
